src/beamlime/constructors/providers.py

```diff
@@ -199,12 +203,54 @@ def __hash__(self) -> int:
     def __repr__(self) -> str:
         return f"Provider({self.call_name}, *{self.args}, **{self.keywords})."
 
+    @classmethod
+    def __copy__(cls, _obj: Provider[Product]) -> Provider[Product]:
+        return cls(_obj)
+
+    def __deepcopy__(self, _: dict[Any, Any]) -> Provider[Product]:
+        from copy import copy
+
+        return copy(self)
+
+
+Constructor = Union[Provider[Product], partial[Product], Callable[..., Product]]
+
+
+def _dummy_call(*args: Any, **kwargs: Any) -> None:
+    ...
+
+
+class CachedProviderCalledWithDifferentArgs(Exception):
+    ...
+
+
+class CachedProvider(Provider[Product]):
+    def __init__(
+        self, _constructor: Constructor[Product], /, *args: Any, **kwargs: Any
+    ) -> None:
+        from functools import lru_cache
+
+        super().__init__(_constructor, *args, **kwargs)
+        self.cached_constructor = lru_cache(maxsize=1)(self._constructor)
+        self.cache_indicator = lru_cache(maxsize=2)(_dummy_call)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Product:
+        self.cache_indicator(*self.args, *args, **self.keywords, **kwargs)
+        if self.cache_indicator.cache_info().currsize > 1:
+            err_msg = (
+                f"CachedProvider {self} was called with "
+                "different arguments from the last call."
+            )
+            raise CachedProviderCalledWithDifferentArgs(err_msg)
```

Comment:
If this function is called 3 times with argument sets
1. A
2. B
3. A

That is, the 2nd time with different arguments than the 1st and the 3rd time with the same as the 1st. Does the 3rd call work or also raise an exception?

Reply:
I just updated it to reset the cache...!